lib/rex/parser/nmap_nokogiri.rb

```diff
@@ -1,404 +1,472 @@
+                         when 'address'
+                           record_address(attrs)
+                         when 'osclass'
+                           record_host_osclass(attrs)
+                         when 'osmatch'
+                           record_host_osmatch(attrs)
+                         when 'uptime'
+                           record_host_uptime(attrs)
+                         when 'hostname'
+                           record_hostname(attrs)
+                         when 'port'
+                           record_port(attrs)
+                         when 'state'
+                           record_port_state(attrs)
+                         when 'service'
+                           record_port_service(attrs)
+                         when 'script'
+                           record_port_script(attrs)
+                           record_host_script(attrs)
+                           record_vuln_script(attrs)
+                           # Ignoring post scripts completely
+                         when 'table'
+                           record_vuln_table(attrs)
+                         when 'elem'
+                           record_vuln_values(attrs)
+                         when 'trace'
+                           record_host_trace(attrs)
+                         when 'hop'
+                           record_host_hop(attrs)
+                         end
+                       end
+
+                       # When we exit a tag, this is triggered.
+                       def end_element(name = nil)
+                         block = @block
+                         case name
+                         when 'os'
+                           collect_os_data
+                           @state[:os] = {}
+                         when 'port'
+                           collect_port_data
+                           @state[:port] = {}
+                         when 'host' # Roll everything up now
+                           collect_host_data
+                           host_object = report_host(&block)
+                           if host_object
+                             db.report_import_note(@args[:workspace], host_object)
+                             report_services(host_object, &block)
+                             report_fingerprint(host_object)
+                             report_uptime(host_object)
+                             report_traceroute(host_object)
+                           end
+                           @state.delete_if { |k| k != :current_tag }
+                           @report_data = { workspace: @args[:workspace] }
+                         when 'script'
+                           report_vulns
+                         end
+                         @state[:current_tag].delete name
+                       end
+
+                       def characters(text)
+                         @state[:add_characters] << text if @state[:add_characters]
+                         @state.delete(:add_characters)
```

Comment:
This line doesn't make sense to me when I test this out myself:

```
2.6.6 :024 > oldHash
 => {:accuracy=>50, :value=>"vuln2"}
2.6.6 :027 > oldHash[:value] << "test"
 => "vuln2test"
2.6.6 :028 > oldHash.delete(:value)
 => "vuln2test"
2.6.6 :029 > oldHash
 => {:accuracy=>50}
2.6.6 :030 >
```

According to the test above we are just updating a hashes value here and then immediately deleting it, which seems contradictory to the purpose of adding new characters to this string.

Reply:
the line isn't adding a key/value pair to the hash, it's appending data to a list stored within the hash, so when we then remove that item from the hash the data we added is still in the list and is still accessible and used elsewhere

Reply:
Honestly I'm failing to see how this is the case given the above example which seems to suggest this code would be appending to the list and then subsequently deleting it. Can you provide a code example with output that proves otherwise?